src/metabase/db/metadata_queries.clj

```diff
@@ -27,12 +27,52 @@
       :data
       :rows))
 
+(defn- partition-field->filter-form
+  "Given a partition field, returns the default value can be used to query."
+  [field]
+  (let [field-form [:field (:id field) nil]]
+    (condp #(isa? %2 %1) (:base_type field)
+      :type/Number   [:> field-form -9223372036854775808]
+      :type/Date     [:> field-form "0001-01-01"]
+      :type/DateTime [:> field-form "0001-01-01 00:00:00"])))
+
+(defn- query-with-default-partitioned-field-filter
+  [query table-id]
+  (let [;; In bigquery, range or datetime partitioned table can have only one partioned field,
+        ;; Ingestion time partitioned table can use either _PARTITIONDATE or _PARTITIONTIME as
+        ;; partitioned field
+        partition-field (or (t2/select-one :model/Field
+                                           :table_id table-id
+                                           :database_partitioned true
+                                           :active true
+                                           ;; prefer _PARTITIONDATE over _PARTITIONTIME for ingestion time query
+                                           {:order-by [[:name :asc]]})
+                            (throw (ex-info (format "No partitioned field found for table: %d" table-id)
+                                            {:table_id table-id})))
+        filter-form     (partition-field->filter-form partition-field)]
+    (update query :filter (fn [existing-filter]
+                            (if (some? existing-filter)
+                              [:and existing-filter filter-form]
+                              filter-form)))))
```

Comment:
This code modifies a query, which should technically be done with MLv2, rather than being coupled to the query structure directly. It's probably okay to ship this because we already do it everywhere, and I'm not sure if there's a precedent for modifying queries using MLv2 on the BE yet. But would be huge if you could find the appropriate function in `metabase.lib.core` and use it here.

Reply:
I've asked about this here https://metaboat.slack.com/archives/CKZEMT1MJ/p1704972904847889

Reply:
Yeah it would definitely be good to use MLv2 for query manipulation like this going forward, for adding another filter you can just use `(lib/filter <query> <new filter>)` and it actually takes care this sort of stuff for you automatically. The benefit of using MLv2 everywhere is that we get to dogfood it a bit more, people get familiar with it, and we can have one big set of util functions to use everywhere instead of having to reinvent the wheel a bunch of times. Oh you also you don't need to know MBQL internals to use it.